### DISCUSSION TITLE
Graceful Shutdown

### DISCUSSION THREAD

**Discussion by Acewiza** [Security-Onion-Solutions/securityonion#6070] [Unsupported Versions] [answered]
I'm gonna take a wild-assed guess and assume it's a bad idea to issue a random shutdown or reboot command on this thing.  Without a firm understanding of the myriad internal data flows, is there a recommended or preferred order of stopping onion processes for this purpose?

**Comment by Sahale** [comment]
I'm not sure for 100%6 but simple `# poweroff` is quite safe. At least it works for me without problems.

**Reply by Acewiza** [reply]
You sound like the guy on an automotive forum encouraging a poster having trouble speed shifting into 3rd gear to just rev it up higher.

**Reply by Sahale** [reply]
This is just my experience, I will be glad to know if there is better way :)

**Reply by Sahale** [reply]
I tested scenario of disabling firstly logstash and then redis (it was empty) on heavy-nodes and found out that filebeat waits logstash to come back online, then sends accumulated data for the period it was offline.
According to this, I think that it should be a good way to start reboot or shutdown with these steps:

1. Stop Logstash;
2. Wait Redis queue to become empty, then stop it;
3. Stop other containers;
4. Reboot or shutdown.

There is an interesting moment with heavy nodes, which is not described now in docs: logstash sends data to redis, than get this data from redis again and sends to elastic. I'll try to edit docs to describe that.
Hope this helps better than previous answer :)

**Reply by Acewiza** [reply]
Yes, makes sense.  From what little I've gleaned, logstash is a good candidate for first or early stop.  On the contrary, what should be waiting for later, or last?  I count 21 candidates eligible to receive an so-******-stop command.

**Reply by Acewiza** [reply]
Database inconsistencies are never fun....

**Reply by Acewiza** [reply]
It would be nice if there was only 1 or 2 things you just stop them 1st, you'll never have a problem related to bouncing the box.

**Comment by abesinger** [comment]
I've done numerous reboots without noticing anything obvious. I haven't done any power-downs.  The SO architecture seems to be pretty robust in that manner. But if you wanted to play it safe, I'd follow the flow of data.  Shut down sensor nodes first, then the manager, and then search nodes.  That way there's nothing try to send/queue data without anything to receive/dequeue it.

For startup, I'd start the manager, then search nodes, so that the manager is up when the search nodes try to connect to redis. And then once those are up, start up sensor nodes.  That way everything is ready to receive data before anything gets sent.

If you really want to get fancy, look at stopping salt before shutting down, so that you don't shutdown in the middle of a highstate call.

**Reply by Acewiza** [reply]
That's a good way to think about it, but I was thinking more in terms of just the data flow between processes within an active manager itself.

**Comment by TOoSmOotH** [comment, marked as answer]
If you are running standalone then normal reboot or normal shutdowns should be fine. If you are in a fully distributed grid and you want to bounce the entire grid I would do the following:

1. stop salt-minion service on the manager.
2. run so-logstash-stop on the manager.
3. keep running so-redis-count on the manager until it is 0.
4. reboot the manager
5. reboot everything else

**Reply by Acewiza** [reply]
Perfect - tks, smooth.  Sorta reflects what sahale was saying.  Maybe a little background for why I ask...

I ran SOv2 for a month or so when it was 1st released, but it was somehow incompatible with my hardware at the time.  After suffering repeated random lockups, it got to a point where services were failing to start.  I worked through that for a few days before realizing it wasn't a software issue at all, and by then the Onion was utterly failed as well.  Went back to ubuntu/16, no problem.

So it IS an issue, I just happened to force it in an unusual corner case.  My new onion is happy!  :-)

**Reply by Acewiza** [reply]
The other thing that actually prompted me to make this post was when I realized there is no so-all-stop script corresponding to the old so-stop-all.  Remembering what I just went through, I'm sitting here scratching my head thinking, OK let's see if we can screw this up again...
